pyteal/ast/abi/util_test.py

```diff
@@ -0,0 +1,274 @@
+        ),
+        SubstringTest(
+            startIndex=Int(4),
+            endIndex=Int(5),
+            length=None,
+            expected=Substring(encoded, Int(4), Int(5)),
+        ),
+        SubstringTest(
+            startIndex=Int(4), endIndex=Int(5), length=Int(6), expected=TealInputError
+        ),
+    ]
+
+    for i, test in enumerate(tests):
+        if not isinstance(test.expected, Expr):
+            with pytest.raises(test.expected):
+                substringForDecoding(
+                    encoded,
+                    startIndex=test.startIndex,
+                    endIndex=test.endIndex,
+                    length=test.length,
+                )
+            continue
+
+        expr = substringForDecoding(
+            encoded,
+            startIndex=test.startIndex,
+            endIndex=test.endIndex,
+            length=test.length,
+        )
+        assert expr.type_of() == TealType.bytes
+        assert not expr.has_return()
+
+        expected, _ = cast(Expr, test.expected).__teal__(options)
+        expected.addIncoming()
+        expected = TealBlock.NormalizeBlocks(expected)
+
+        actual, _ = expr.__teal__(options)
+        actual.addIncoming()
+        actual = TealBlock.NormalizeBlocks(actual)
+
+        with TealComponent.Context.ignoreExprEquality():
+            assert actual == expected, "Test at index {} failed".format(i)
+
+
+def test_int_literal_from_annotation():
+    class IntAnnotationTest(NamedTuple):
+        annotation: Any
+        expected: Union[int, Any]
+
+    tests: List[IntAnnotationTest] = [
+        IntAnnotationTest(annotation=Literal[0], expected=0),
+        IntAnnotationTest(annotation=Literal[1], expected=1),
+        IntAnnotationTest(annotation=Literal[10], expected=10),
+        # In Python 3.8, Literal[True] == Litearl[1], so the below test fails.
+        # It's not crucial, so I've commented it out until we no longer support 3.8
+        # IntAnnotationTest(annotation=Literal[True], expected=TypeError),
+        IntAnnotationTest(annotation=Literal["test"], expected=TypeError),
+        IntAnnotationTest(annotation=Literal[b"test"], expected=TypeError),
+        IntAnnotationTest(annotation=Literal[None], expected=TypeError),
+        IntAnnotationTest(annotation=Literal[0, 1], expected=TypeError),
+        IntAnnotationTest(annotation=Literal, expected=TypeError),
+    ]
+
+    for i, test in enumerate(tests):
+        if type(test.expected) is not int:
+            with pytest.raises(test.expected):
+                int_literal_from_annotation(test.annotation)
+            continue
+
+        actual = int_literal_from_annotation(test.annotation)
+        assert actual == test.expected, "Test at index {} failed".format(i)
+
+
+def test_type_spec_from_annotation():
+    class TypeAnnotationTest(NamedTuple):
+        annotation: Any
+        expected: Union[abi.TypeSpec, Any]
+
+    tests: List[TypeAnnotationTest] = [
```

Comment:
@jasonpaulos I'm wondering if there's any reasonably straightforward way for the test to catch `TypeSpec` additions unsupported by `type_spec_from_annotation`?

* I'm imagining it's possible we'll forget to update `type_spec_from_annotation` when adding new types.
* Here's a couple of avenues I briefly inspected.  Though I didn't go far enough to be confident a palatable solution exists.  If you feel a path is fruitful, I can do more homework.
  * Quick googling suggests `__subclasses__` might be helpful (e.g. https://stackoverflow.com/questions/3862310/how-to-find-all-the-subclasses-of-a-class-given-its-name).  Some thought required about how to apply the knowledge to the problem at-hand.
  * Python 3.10 pattern matching + mypy might be a way forward too.  Via https://github.com/python/mypy/issues/12021, I found https://github.com/python/mypy/issues/12010.  https://github.com/python/mypy/issues/12010 suggests mypy can exhaustively pattern match. 

----


Aside:  Do we expect PyTeal developers will extend `BaseType`?

* Obviously, if someone extends `BaseType` without modifying `type_spec_from_annotation`, then the new type can't be resolved.
* I realize `type_spec_from_annotation` isn't used yet.  I'm assuming the intent is to start using it in https://github.com/algorand/pyteal/pull/170.  If that's the case, I'm trying to consider how relevant it is to support arbitrary resolution.
* I'm not looking to overcomplicate what's been done.  Though I want to confirm we're _not_ missing an obvious concern.  
  * It might be sufficient to defer the question and/or add a disclaimer in user guide docs.  Even if we agreed to take action, I'd propose doing so in a subsequent PR.
   * Another direction might be to disallow subclasses.  I'm not sure if it's possible in Python.  On the surface, it doesn't feel to me like a typical maneuver.  Quick googling suggests it's _possible_:  https://stackoverflow.com/questions/16056574/how-does-python-prevent-a-class-from-being-subclassed.
  * My starting assumption is that we _don't_ need to support the concern today though a user guide and/or inline disclaimer may make sense.

Reply:
Using `__subclasses__` to check that `type_spec_from_annotation` supports every type is a good suggestion. I've added a unit test which does this in commit 71a1040a6657bbcc43f5d54d5eb06939fc13eb4c. This should make sure we don't forget to modify `type_spec_from_annotation` when adding new types.

In regards to the second point, I expect PyTeal developers to _not_ subclass `BaseType`, though it's possible someone will want to do so for reasons specific to their use cases. And yes, in this event `type_spec_from_annotation` would need to be modified to support any new types. There are some things we could do to make this process easier (perhaps some sort of registration process that allows the function to be aware of 3rd party types), but since it's a use case we're not actively encouraging, I'd rather wait until it becomes a problem instead of trying to add complexity we're not sure will be needed.